include/up-cpp/utils/SafeMap.h

```diff
@@ -0,0 +1,336 @@
+    using reference = typename MapT::reference;
+    using const_reference = typename MapT::const_reference;
+    using pointer = typename MapT::pointer;
+    using const_pointer = typename MapT::const_pointer;
+    //using iterator = typename MapT::iterator;
+    //using const_iterator = typename MapT::const_iterator;
+    //using reverse_iterator = typename MapT::reverse_iterator;
+    //using const_reverse_iterator = typename MapT::const_reverse_iterator;
+    using node_type = typename MapT::node_type;
+    //using insert_return_type = typename MapT::insert_return_type;
+    ////TODO Base-class-specific using statements (e.g. hasher)
+
+    // Pass through the base map class constructors to allow construction
+    // from an unsafe map. Also covers some of the generic iterator
+    // constructors.
+    using MapT::MapT;
+
+    // Define the constructors we need for copy and move
+    BaseSafeMap() = default;
+
+    BaseSafeMap(const SafeMapT &other)
+    : CMCLock(LockMode::Shared, other.map_lock_), MapT(other) {
+        CMCLock::unlock();
+    }
+
+    BaseSafeMap(const SafeMapT &other, const allocator_type &alloc)
+    : CMCLock(LockMode::Shared, other.map_lock_), MapT(other, alloc) {
+        CMCLock::unlock();
+    }
+
+    BaseSafeMap(SafeMapT &&other)
+    : CMCLock(LockMode::Unique, other.map_lock_), MapT(std::move(other)) {
+        CMCLock::unlock();
+    }
+
+    BaseSafeMap(SafeMapT &&other, const allocator_type &alloc)
+    : CMCLock(LockMode::Unique, other.map_lock_), MapT(std::move(other), alloc) {
+        CMCLock::unlock();
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    /////////// Locking wrappers for existing std::map interfaces /////////////
+    ///////////////////////////////////////////////////////////////////////////
+    SafeMapT& operator=(const SafeMapT &other) {
+        std::scoped_lock lock(map_lock_, other.map_lock_);
+        MapT::operator=(other);
+        return *this;
+    }
+
+    SafeMapT& operator=(SafeMapT &&other) noexcept {
+        std::scoped_lock lock(map_lock_, other.map_lock_);
+        MapT::operator=(std::move(other));
+        return *this;
+    }
+
+    SafeMapT& operator=(std::initializer_list<value_type> ilist) {
+        std::unique_lock lock(map_lock_);
+        MapT::operator=(ilist);
+        return *this;
+    }
+
+    // skip get_allocator()
+
+    mapped_type& at(const key_type &key) {
+        std::unique_lock lock(map_lock_);
+        return MapT::at(key);
+    }
+
+    const mapped_type& at(const key_type &key) const {
+        std::shared_lock lock(map_lock_);
+        return MapT::at(key);
+    }
+
+    mapped_type& operator[](const key_type &key) {
+        std::unique_lock lock(map_lock_);
+        return MapT::operator[](key);
+    }
+
+    mapped_type& operator[](key_type &&key) {
+        std::unique_lock lock(map_lock_);
+        return MapT::operator[](std::move(key));
+    }
```

Comment:
Looking over this again, I realized that returning references is probably not safe here. Two reasons:

1. Since the lock is not being held, the reference value could be altered by another thread
2. Another thread could _remove_ the referenced value, resulting in accessing invalid memory

I put together [a demo](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMRMAPVgB3oDpkPHqSqoGBAMrIm9EAEYALKPEEAqgGdMABU4gA5AAZ9pAFYLStJg1DIApACYAQo6ekt9ZATzjKjdADCqLQAriwMIACspP4AMngMmAByYQBGmMQgAMwAbKQ8qBqEPgxBoeFR%2BYXF4vGJKSzpmbnumJ7e4pIETMQEZWER0R6YXiVdPQR1yWkZ2Xka3b39FUOLkwnTjbO5AJTuqCHEyOz6jlkJyKFYANSODpJMVJgAskw8/Ah3doYAgt9/DnODEuIRudiyAR8C2ImFY4Ow/zOFyumFuEOhCWA8MRv1SqGC1xY4kIJDRABFrlRZFpwS5fv9oSEvNdnsSiDD0F1iJjriBrjwQqlaHhkNcFugQCAMTZbgB2Ok/a5K64hIoy8WS6WgKUEbk2Wk4xXKgBueF6IVk1wAfqyGCSOVzMRAdnKFcr3dc8FRrhAiXb2S67PL/h7Q2KCBKUAcCGiAuCArcHA4bWySJhObqnTs%2BWBTkmDb8wx6NVGQjH4/HrgAqAgIPAaWOVu65n1fLJuovKktoMuNiHXWv1vsJu47O7D8OR/y0AtG0NBskh5ULxGyxftw0JGMsJgJZ2updK1U8h5PV48ckqnjEVBENC0SVlugaSWnl5veNb0gs1MOzP6rIEQ3QtlTfc9CQ/YN5Xkb87gACTaWhUC%2BNdYPlBxYKTAB1EhaHQFCyTQpwskwhwwDAAiV2Av4QKVbtownO4dx4OxIicBxWIpcF1y4pMJ2Y1j2M4icS2nWd3Xo3sK37O4AjeAhDh5GEnhhYFUSICCWLYjjIkXPjpITUSGHQGdqPdJgy2QhwckpfFLwE7TOPErsI0lHtywhJskwAeUOa5lIyRhjmuNAJF3BgX0TBwJzEVBGL4iBxwMxMcli%2BKHDHfTPP7IyTNnQ9NP4XVrA0JgvEStjOIgMC3kfCLHkwAB9ZjGoIABPHhUUcHIWEDYNaI9P17UvXUQkwZzQ0kjy4xkpNsGIJgtE0wSdL06Lkty0yO1DFh%2BAyRbMAgDKJsG39LypWgaTMsMpvSn4NCKYA7RlDSHKcRQnKymbDNcjhjK2gr3RYQSPt0y87mwH4ACUABU4Lbba5TJMdqIK27kruXziH8zAVKC1FQu6BJIqS7KEzSjGEtJ76Uopsmosy9b6c2/K130PYH30SITAiIwTFQDhuBYPhMEEYRSAIAxjD2CAkDQLBcHtchKBoehmDYEAaveBAVAkaRZHYJRdfULRdClkxzAwqwbBQVxXFadoSj8YzlgiayYmMqYGiabIsgATiqIoOlKYIBhAd2CiDkovZmZp/YdkZg/GJZQ4qd3hlGTo1hjrY44DhYJld8P5mzjZve2f29g0A4jhOAB6KsQyra4gh4NruWABAYwgZAXQcQwHEUa4AHFGH22gf3ZBth5h7zrliWI40LRvl/nkVGC0dABzi2tUR%2BHgyoQVFJFQKgCAAdx6VEADEDmMphg59H5JGvl0QmMjJrnEcbV9JIkYRCqobkqRLLY2FMcCKqImDABhJgNgEgND8CVJITA6kj7XCSN5GGABJAI2BKR0B/oqZu%2BAMQgIIOma459CAIAHHWBs58SAAGtbLYyYOgfAwdLQJDEMQHcwcm442AD0fAMo0Btw7l3L%2B59EjEA0HWd4SoYboOftfa44CN6YAbIOBsqsiHVm3tcNqBwVQf2xrva4%2B9D6oniBArQ34ABqGQijiGuA4fghgfS70EXcWxmjRy0iMSYncbVrgMFvCqJaOiCH0GuFwY4PAYwJEAcLYU1hjiCOobWOhNj16QKQdcAAmsEpgoTUCpCJgwa4TBAFty/t6CxfjIHVIIE3QRSou4EB4CAOuddz79P4G8ax/ASDADrhoyBGg66xFwdgJIkhsAAFp3GGDaavNQDB6APRxgARxCGaShqRQlvD4CKJgQpURWHPl/NhMDUFbw0sk8%2B3JvA2FIII6uZ9L4ANIZmchlD35YHMegppS0hy/OAWWShrjrA%2BKTM/a42DJDjicM/JF34sLYLht5NQMNrhYWhlDH4SQcHYEkNcbyUMW7eSSGSLF2CaXku8mo4lhTBEAGlsG0u/JgGhn8uDXi0Q2UkeBUl4HTAUlBaDcl2NRLwnJHzOrIC9CKdR1hgAWmAKiYAqBjQZGesAfkGQWD1hcRFapxl1GisIA/EoGhBGAs/o0vJWh%2BCCMkNoMkAANRZ19CEww6uwa4kgcVQzwe6z1Pq/X0Fbu3PAncCAwy4JLNxA8h6j1kZaVkU8R6z3novCN3rFmgswIs7BWAJAqtmJYg%2ByAj7LI8U3OuiJAReg/t6NQ2hGoBG0F2yQPxr7YEas8H4Xb4b0kBFgGgiRridu7b2xq/bB3DtHY1cdNEARAhBGCCEVB36Z2YKZBEE6t0oljAJQCLbT2gi6hCFgULODYhPciG9sZ5FX3QM1B9T7N0vp3QEd%2BJAgXpmah%2BS99Jfjqy0bW1EIRry3lQPeR83hLoHggz8ChqSH7jQhO1TqUGWRvBhj%2By4i0GxOAOlrfKvxMN8Gw/GUj2z2WYDat%2BRjDYYbvIGuxluqBhZXxGoGgjJYL0BGY6x64xHAKSiYSxxqaB%2BMwi43OEKVhtk/FoEhGQ7JBP4dYKiETYGxMse/Jx3jimcPYElLILTD8SCtUDT%2B48MotaXgo1oKj6JfqifE6Z78QQLPfg07Z9k2JL2oxo3AujFCGNqYbL51TZHJPKfdDxuCi1aHcQHEJ/Tk46pAcCp%2BnzJnJPwklAgDLGQUvKh4%2BJ7AezLRZbw4wXLJZAPEGA0VozCWzPpfkWVkAsm2qNUwA12g1WlQ8eC4huz2Mms5bYHlkA7XOugZYhCHr34%2Bs62uHVsbA2bMzfZA5zqTm1SGrfBsgrHJwJZfc5gTzAQ2vhI64Vtb8ZNvXG29%2BPbFpxuWM00dkgYWgIKn%2BL0uuiKJD6stFgImqH5X3dc6KkW8DugCMg/pjQMHrhw93BPIMCoIfXAAAKpG5LjSxVSQjAFoKEiryAWEaS0ywh%2Bnp/R4FkHgAAXra1x3hFvoEUmI1AdTrBbyJHqwBEVRoHvB70%2BXkPNCUOZ4hpn6C7i3iPsQcc5TTCJyoUfKp4i2qYjrpLzESCoajf2TCBs1g4mcFOcqmMMhNOK8Me/Fn%2B5HgUOBaiBIxQufc8/sKBYnoGzoG/j6ZJFivCkjxOgNqOw3W/EZMyWNrI9UBHZLENXn9CcFUz7qzAOeSB58ZxkfcWWp0WVoK0iLKnGBhES9sivTDWRgn6ip90SRv7fg2XgPZmBvySAqxyAqVEwcDWL9n3P%2BfiAQHb531ERIsDfhLO%2BjkX6KGcBSr1MJqB4mYESdRlTy%2BMCwc2Wr/c4Tj%2Bn8b9ePAxpsMgAKhfm4yAWCXDV81S/hOP92AQA%2B9Egp80Zfp34h8xpGoWd4xN9x8QN71d94QVRYDro6Jfot8QM0CntMCECitv1AIxQ0Cwc1xWYKRekwksccc8c6BDRicycKdvQsIFoeAGxDMLxkkal4CP0d8uBWE6EYR2FFlSonhqlkBjgHpFcPcnBQlnNDUeDt8kCuBvxz4CYHcmBjRUA8At4tBuRg8%2BcqlT5pdw8ypJDtFkJfhicLFmIkEsI6xYld4AEhwFg6AJ5Gpq42BGojViBeEdw1Jrg6wYwqAbwWBqkOEeQLEWdlNrD6FCRihhEKFtF0FGE5EYxjDCBU8fgIcPd7odC5VQicl1E1dMRvwbN2c/cqRJDqkAED4HpawbwacEALCijHViA6dMQPcODW9EFpCrDKCScYQdxiAmEGwsJYMlp4MKEvBKEEhtcbUAiQi%2BMltmIxQd4KsYxrBjkCBaMYwiAPdOxVQoEqktwMgqi0E2cKsGxwkYx0hGB%2BQbxpiKEt4jlqlCQH0qE3DKR8cDiiwNIFMeBCFPjskHcMgbxgU2cbDMBawMAI8HjbxE4JU%2BjsiBihiehRiKUqkLENBBR9dmQMi/dZt7dLVip6oD0Xx4xycf1idOxlEhxnkTlP4qwo8tFD8CBm5Dtrl8AYRmQzChVDFTjfCypWTawH5fiwxuhZMbkcYFJiATjCSp4kE6SGwhx0gZAjjPQFSSB7daAhCk8QpYV%2BjIdOx0g4lFo6BQkEgX9hR0BsMXjQl/CylOoFop4N84p2FRFDVQpkBDhVJWkjTOxxDzCIJZMVSYxEgeg6dPRhZCgigLlPcokj5dp8UHDLkSiRcQg8JxTQxTSPSVd1iA8tS5EckWAPA9UNBvxRSYxGFMyt4twbwhdgpEhz5syPRuQNAMSIBRZgAkEWd7cJD8iwpNNQl0geQj48Jrg1SLIloHdCB9pQsAzOxL4Gxxz0AU9Wz3RKSddwMUTjSixlEDNQ5H4hwNJ4NTQbhSTSpyoXQ2AYT0Bki2dOSBCQFaAmENzlQypg4GFUzAjghPSiiWckFb5sZRd5y7VDdApNT5ztT3ylRSQ7MTVRRQKXTwLEh8yv4JC/JXFVjbTugyiDTNNzl6BYKfQ91gRH4ChBTvwrBGhbSeUCBe4DSqlTS6jN5DFLzPznQlSj4SKWdPjNNJzURVzDcgS48bMiKA8GxiB34DUsjaMrAYtcMFtURzwpNj0fgeMkcGpwI%2BQn8X8KFcdoT8dJRZ9S959K8yB4SnjKFVK0MVMkCJKlssCCDd9NIYDf9C8yD0DIkeQNlRD3weA8UstVKTp5Dg1tKiM3NKMIr1sAgQrwtEYwq4qAh29LxaCHwQBTKy9iB28Mgz8BQhQRQ38BoIdSqyryqKrKqqrqqaraqCparickhMBrlBTzjdFSR/LJzMymcFoySvzrgGrSr6rBqRrRqxq6qBp5L6MlK9NFtYYf13QwrO8vQ2oYZOAqkssSwyK5cIRYYIAQrrIUZQdDR3QprFKc9lLrh5qdz3RYYBxeqryCB9qMAVq1qGB4xrr8FupuoqA%2BpEYJIIC7Rh93LGdijGdfQ3gQamFGoUZ/rlQYRZSqkqAIAax6FYbJ8yD0N3QtCdD7qSpOLlqqBVr1r4wcb8IiDvrrJfq7K/iHrPzSbtDybsAIARMS9qAdh0aBoVwsblQzqcMLrZrURPqCokrxAFg3rLwtr90BFdqYZ9rIrQoFhuojqz9TqosFL%2BbmsCNhaBo7qOLyoggZc3qPq1KUofqXRFbyxu8ixFDsC1cwamEIaeAoaYaTp3QEbDgkaUadFOaVNuaN1sbGa8a%2BqIBDbxaSaIQyaUDKbUoLaxara4alR9aG8Ago7AIWbvM2bfrfbNzMaA7lRKDtBBRwFstOoBCLEGThBKEBIBowr35/K1sTtb0KQ/KYqpNp8VNxqu7u7xrhqu7rhtAyNFkGiDhO52rsZuie6yq%2B6p7Z656ciSrIdB7tkR6mjqkBKLFmtDLp0ahzUuzOAEk9iEBCgDMlUVVRR9jF6xRCALR%2Brz5tdCzoLiyZAqlId71w9TSjj0AsjFrztrghsm7dMWtFtVKZM5NmtQq/7mJOpP0t75tBbCNArJRoGQMIHvKwrrToC4GsgKQtbctQGQBMGmo0HEq/6igQ9AH4HgGVKiNNQediHHN0G/78AqA8Y1JKGcHS7qHEGYZJQWG2Hjgm6TpekwqAGASBMqGCMCGxG%2BMD4YRhG64MHZBoDxGAFJH8HaHCHlGmpVGcNSGeRDttN7NsHcHLqCHDHZshGmGlJcZAoAj1GQHNGApVIMlrGRcZdGpnH8YgGpHNHLbPHbGXG9GRa/7KLoc5tOG8HHGkGQAwm/dIGeR/G4mC9ImzG/H47GpkntzEYRG/65yXTSQHGaGYn8nZsFGwr/HSmdMimeG3IMmqngdvLcmbG9U5EmoGmInTGEGCGYRWmtBGoOnym/7/HennF2miyfGNGYmRnMA%2Bnxmn6EmZRwksAOGunuGCHlmGHTsmnFG8nIFegAnEbVmuHfGSn9mCBDnParGcmIdZ4yQ557tFl2MRCz6aBRQwqC5MNGACAGwuz%2BAezAjKtiAxwealRC6ktV7O4ijUgDpNIeiTDZcc0NInzLbpK5cr6liwiHd66GpCoAdq5AE%2Bm1jFtjCLEtVZFVVBmr7UWmRFSQmeQCH4qO73RKCyRcYNgiiaX49iy1Cwl7kBCTcLUJcS8CotKzw3hq9OHa9MyG9mXlQxWAqe547wrxXAqUotcMgQWVM%2BRkr28l81cV9Xx8DvwNXiB%2BAWoWcYbvxVKIBTW/rAZlRdW1c6pvcc7lw87RXorVWlWZcVWAqgrrJTW2NlWLHjs4HrJDGtX3QdXngUqb9ACjWP0TWnDzXIbLW9geHbWnCyjAdGLC8Bp3QnXGcXWb83WlR/bPWPMYrqo26zbA2nCo3lQY243waE2QBB9h9k3tdU3nb03rWiMM7IxJdDo7X7WC3HXY328S3way2kYTr5WvXFWtYA2chuog2EX16QtjHA0UpI2Cpm29W22O2xou2Mge2XaM2bXWa9Us3tcM3I2aawwi2mFp3HbZ2K3F756v2p6Z6Kr550zDVK7nTx7Hd6xXkFDvM3gKizjhSGwGrf3v3EORqCpl3uov5nSymcGfXw9l31WG3H3JpMDxE7bQbvcLXf9T2zXyPGdXbvL3QCHkLMOyRb3NW3b4boTPbqwdEToP2VNUPrJ0OFmsPcOY67XD978E6HWMDIwNBiPP0%2BKyO02KOv4U3qPobZ36PNHGPQssPr2R2G2NP2PEauP6EeOPWBp%2BObJtPGnmOSxA9vBg8MgYCwPSbtGrH8E6AwOx2e8XJIxIDgaFOb81PaPE7amQBrPtzmPPOFhDOlQPa5STP6wzPFxQWBrIcOy8ALwtULnQ2SBnR52lQUHYHHMBOH5sOYwAHw2cghtvObbAaoCmpAvwbgvYuZTOPzGnqavkuCuN2iuhHSunrLb/7wGd3uoauN382fPpPNR8CoaHanaL22O4uOOEuOuIAuvvLeOgZGTivTsBOIvBIqohvKvRvrJxvJu6u/OgboCmvHaWulu2vVutOMPQtKpdJ1uWN33zP7Kdv%2BurOXvgc3vmOTuy6Y7zvrabp6uAv7bFPe3f9Wv4uqkGOAedcgfB3kG2aauvuUv86wX0umFMuoKCnizFk6ET6xRj7ayv5NkRyJiVcEBGioXN66bmRWrYPhqxQCeLx0hgA9wXRxcQoee%2BeOeMuLx/AfdLUIF0B8ur7Rf/IheGAJet5kBiAFeZfO78fCfVJpf%2BfJftf1fJ83unBln6wZAOsnIwY8QCQot2p9whu78D6T9JPx3pupRZvbuFv02HvEewube2p8vNvvuJJ6HAHyHDo47fWHfD6CPiw8DeCPf7u6OjP2vNGw%2BA/EYtuuwQ%2Bt6dxOAPD6G7flWo%2BneY%2BAaZP3eYegulOaOEeVukfNHc/8%2BQ90%2BMaceCoybVNYRF8XRi/ElS/fO6oGu5vYfFuk%2BlQCHLgu%2BW%2Buag%2BC7NeLw18q1iyqycYozXEOmGxSetAuokw5eN/xxzkRWr67m55tAbxzyDN0dVVbzj77y6E2cX7BL%2BQyN0LN71roOhTJCOfiYMgnrG28f2cehC5r1S1Tq8WWkOH/gc3sxkZ40ivf/mlziRYZjgYAufogLoyCM6wEgFAQAN1DDZ1awpbAQgP2haBCBlBZNAtGvKT4BoHfDQJfB4Dy01Wa7fDr32d5Tcls/nG7pX2a7V91OD3AhrQIlZ2tuuqXSgnLzYDEAtU1wUntgjoQJAWEhAKhAcAnKpEWE06EfF1RjDYJcwWLS6HFBuJihDgxxYxKASvpDgwqCgyROGVQDnw3SVCTALmAEqXwFB78FDJqThJoUsA%2BEVLmH0AY9gsBx3EbqDzO6fcJuEPQjuX3j5cC7uPAkLlJ0e719pmd8Trp92EG48Bw%2BA86rVgWpZ8KGW9PwYN2VbsoUonACPuHgu6Q8Ih2%2BBPjENr7GcJ%2BSQiACUNSEi8ueBCS1KTwsE24DkYZJ%2BhWTWLvEP6AeKpDUmTolAeuVvCeITHCgaByuw3YbFV3G5Ddyh4QmbpENI5V84eNfb3nXzC5TDiYH3ZPM0MmoZD%2BaWQm6sqAmHS5KkMwobkUO6glDQhoXW2vJyiGe94e2wuoekzCj7CmhgfNvrL1aE25ZAnjdVKiA6FFl7UAIwnkhDULEBGoeIQFEKyvDOl4Rd8LeOCN6EtDCeMjYWEiKIbyZZGUgnJKEkf6ClPywJWhPv3pBeUO6lBE/ikjkb1hXEEXDfJgToHOgyilqYge02RpatKCY%2BJQQ8iPhLRTSKOegGjkoRZJaELOUotKXkSCi7B1wUwKqBjBC5cRK/LcJYRRLB1HqdqcENfENBQZscwpQyvDjspZVzKsmTICAEAK8Y24WeMyuXgXyvtb8R%2BR3n32WGehvQPcb/JazXzN0ssR7a7pgFq5hhMyaufgF7lLYPdwxjOfgDCHoAHRp%2BftNcHEkugB5vRX%2BH/DR39HkhAxBrS/Imw5ChjJogFKMTOwe4aBAKCY2ECQOx4FQLIGkEgPGgSDAjcxWWLMX6P/zeVOxv%2BdsZwzbYgFgmA0H3s2N56Ho/8WAcgoaAtFOiLKJlUXG1AdHZVcqi%2BW0f6NZGVDECHxbqAfhYH/A%2BQvYnMZfl9CX4SxXon0EeOhr9ieI/7RnIa3bbBjzxi1PiptSh6cDmu2PNMWCkzG%2Bi%2Bxl%2BPMQOILFYAix6YZ8V2FfGcNnhUNX0PWOpFt9UxlBI0TQSMp0F4JCA5CSaLgyPFEMwQZDM%2BENBnB/AXoNLnOh7R9oB0Q6EdF2nZiWADA3MUgLzGMCkABYWsD4BLHNgywkAuoVBMrB9FMBZMPwCQS%2BH0CLIyQcVH4OyiHROA1A2CWIGSEagwxCk2gICGSDZYgJgAJgAEoQhyrqpNUmActLbGcD2wfSCwPjL5C6Rlho01DAwCYD0QvhYA6Ye0IZJgh6IXJogQhFBkFi8ABAQgEQKAH0AmB6w2CYECiAlB8xyAGgV4AkBPiHBjg5sSKd5HwwJTrYwAQyYFJlhOT2QhkjCG5IlC5TPJ%2BmTWDFXYkBSgpGgEKdunTAJT6w0UhgLFNrgGB5A5UpKYwBSnqp0p7MPYPWDioCSdAN4PEpLH0DNSzAFgZstZMMlZAOY9EnmAlIFj6AYYsCRMFkBGlDTpYewF9pEEMB0T9AigWaRFPmkmAXw20tabRLgCwAYAiAKMMLB0l8TtJ9ATIMAHkBGxVYfuF8BAFSAJTUgrY4gG1CakmBvpzAX6d5FSCDT/ppABTGjm8g08EpWADSXJE0wiSRpWAfwrThODMTuSicZ/FogSlcARgUKGqdDk5jMThQ5OHoG1CCAgSIpmYFgODNaZ4gtAbLVGcKBth8w9gVAKwMAA0AOJxU58VqbzBGliA9YbuQ2MoCFkmwdAegCKZbEsDqojJLgZwJYDwCpAHJHI0XF%2BVsksTWm3ILAKrKrhtAsZvgLsi7FTgRAYIcQMuLHAsCRxd6RcGCDbODg5wfYMEDOEnDWB2yE4B6ZOOsHqBWyXZ7s02RYALi9AnZsweQHsB4mGxppXMfacxPmnXBOAAADhyCLIcgQ8YABIWuDPT%2BAQ8CAItNQTLTw5JgU6RzVICbTtpnMPaaQFplbTGJc0zWS%2BBgglzSA50q6dGAFAEA7psjHSUkCKn6Ak5KctOdcAzmih5AWQfgA4DsmqB9YcgF6aoGVxmxpZ4cCWLAh0IWBK6/06OQxKYn8wDAFkjufUgTnJzU56czOdnNzkBZASD05aRlGLmcS9gR8dhLMA5GVyTANc7aTvJYkNyQAJ0%2B%2BWXKiAVyDAWQWObvMCkcS2Z0cyeXXIOmayS5ewPpk7EUBAA%3D) that shows a reference to `map[2]` that ends up pointing to `map[4]` after the map is updated.

I really see only two ways to deal with this:

1. Return only copies of map data. That would allow these operations to be used without altering code, but breaks normal assumptions about map behavior.
2. Only allow access to these interfaces from within `transact()`. This keeps assumptions unbroken, but requires changes to most code that interacts with the map.

Reply:
For reference, this is the output I got running the linked demo:
![image](https://github.com/eclipse-uprotocol/up-cpp/assets/10570177/3e19e8bd-b036-4d45-bb80-c660063f9750)